app/uk/gov/hmrc/pillar2/service/package.scala

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.pillar2
+
+import play.api.libs.json.{JsError, JsSuccess}
+import uk.gov.hmrc.http.HttpResponse
+import uk.gov.hmrc.pillar2.models.errors._
+import uk.gov.hmrc.pillar2.models.hip.{ApiFailureResponse, ApiSuccessResponse}
+
+import scala.concurrent.Future
+import scala.util.{Failure, Success, Try}
+
+package object service {
+
+  private[service] def convertToApiResult(response: HttpResponse): Future[ApiSuccessResponse] =
+    response.status match {
+      case 201 =>
+        // we're using try because HttpResponse.json is not a pure function and can throw an exception
+        Try(response.json.validate[ApiSuccessResponse]) match {
+          case Success(JsSuccess(success, _)) => Future.successful(success)
+          case Success(JsError(error))        => Future.failed(InvalidJsonError(error.toString))
+          case Failure(exception)             => Future.failed(InvalidJsonError(exception.getMessage))
+        }
+      case 422 =>
+        Try(response.json.validate[ApiFailureResponse]) match {
```

Comment:
I'm not sure you need the Try around this. The validate should simply return `JsSuccess` or `JsError`

Reply:
just seen your comment. This is frustrating

Reply:
unfortunately the json method throws an exception if the json is invalid:
```
[info]   - should throw InvalidJsonError for malformed success response *** FAILED ***
[info]     TestFailedException was thrown during property evaluation.
[info]       Message: Expected exception uk.gov.hmrc.pillar2.models.errors.InvalidJsonError to be thrown, but com.fasterxml.jackson.core.JsonParseException was thrown
[info]       Location: (UKTaxReturnServiceSpec.scala:77)
[info]       Occurred when passed generated values (
[info]         arg0 = UktrSubmissionNilReturn(2034-11-28,2034-12-28,false,false,LiabilityNilReturn(NIL_RETURN))
[info]       )
```